### PR TITLE
feat(cli): add --file filter and multi-source support to argocd app list

### LINKED ISSUE #21052: CLI app list should be able to filter based on a list of files
# Summary

Argo CLI has multiple filters on the `argocd app list` command. However, none are allowing us to filter based on the source `path`. This is very useful for monorepos when filtering on the --repo returns too many applications.

```
Flags:
  -N, --app-namespace string   Only list applications in namespace
  -c, --cluster string         List apps by cluster name or url
  -h, --help                   help for list
  -o, --output string          Output format. One of: wide|name|json|yaml (default "wide")
  -p, --project stringArray    Filter by project name
  -r, --repo string            List apps by source repo URL
  -l, --selector string        List apps by label. Supports '=', '==', '!=', in, notin, exists & not exists. Matching apps must satisfy all of the specified label constraints.
```

# Motivation

The main use case would be to find the affected applications in a repository based on a list of changed files. 

# Proposal

Adding two new filters. The first one would be `--path string` and it would return all applications that are syncing this path. For instance, if the argument `--path example/apps/foo/chart` is provided, then `app1` would be returned.
```
APP			PATH
app1		example/apps/foo/chart
app2		example/apps/foo
app3		example/other
```

The second parameter would be `-f, --file stringArray`. When provided, the server will return any applications that are affected by the files, by evaluating the path and the `argocd.argoproj.io/manifest-generate-paths` annotation against each files. For instance, if the arguments `--file example/apps/foo/base.yaml --file example/other/manifest.yaml` are provided, `app1` would be returned because `example/apps/foo/base.yaml` is part of `example/apps/foo/chart/..`, `app2` because `example/apps/foo/base.yaml` is part of `example/apps/foo` and `app4` because `example/other/folder/manifest.yaml` is part of `example/other/.`.
```
APP			PATH							ANNOTATION
app1		example/apps/foo/chart			argocd.argoproj.io/manifest-generate-paths: .;..
app2		example/apps/foo				argocd.argoproj.io/manifest-generate-paths: .
app3		example/other						
app4		example/other					argocd.argoproj.io/manifest-generate-paths: .
```

### PR DESCRIPTION
**Pull request by Prayasvadlakonda**
Fixes #21052

## Summary

Adds `--file` / `-f` flag to `argocd app list` so users can filter applications affected by a given set of changed files. This is the missing piece for monorepo CI/CD workflows where `--repo` returns too many applications.

## Changes

- **`util/argo/argo.go`** — Added `FilterByFiles`: filters apps by evaluating each source's path and `argocd.argoproj.io/manifest-generate-paths` annotation against the provided files, using the same `util/app/path` logic already used by the webhook refresh optimisation. Sources with no path and no annotation (e.g. Helm chart-only sources) are explicitly skipped to avoid false-positive matches. Updated `FilterByPath` to iterate `GetSources()` so multi-source applications are matched when any of their sources has the given path.
- **`cmd/argocd/commands/app.go`** — Wired `--file` / `-f` (`StringArrayVar`) into `NewApplicationListCommand`.
- **`util/argo/argo_test.go`** — Added `TestFilterByFiles` covering all six cases requested in the review of #27435, and a multi-source sub-case for `TestFilterByPath`.

## Test coverage

`TestFilterByFiles`:
1. Empty files list returns all apps
2. Matches app with matching path
3. Matches app with `manifest-generate-paths` annotation
4. Does not match unrelated app
5. Multi-source app matched by one matching source
6. Helm chart app with empty path and no annotation is not matched

`TestFilterByPath`:
- Multi-source app with matching source (new)

## Example usage

```bash
# find all apps affected by changes to a specific file
argocd app list -f apps/myservice/deployment.yaml

# chain multiple changed files (e.g. from git diff --name-only)
argocd app list \
  -f apps/frontend/values.yaml \
  -f apps/backend/kustomization.yaml
```

## Notes

This is a fresh implementation addressing the review feedback left on #27435 (missing unit tests, multi-source `FilterByPath`). The `FilterByFiles` logic intentionally avoids `AppFilesHaveChanged([], files)` for chart-only sources since that function returns `true` for empty `refreshPaths`, which would incorrectly match every Helm chart app against every file change.

### CODE REVIEW

**Review by nitishfy — APPROVED** [review]

**Review by nitishfy — CHANGES_REQUESTED** [review]
Please fix the failing checks.

**Comment by Prayasvadlakonda** [comment]
Fixed — added `Signed-off-by` to the commit and regenerated the CLI docs. Both failing checks should now pass.

File: util/argo/argo.go [hunk 1]
```diff
@@ -177,15 +178,46 @@ func FilterByRepoP(apps []*argoappv1.Application, repo string) []*argoappv1.Appl
...
+// It evaluates each source's path and the manifest-generate-paths annotation.
+// Sources with no path and no annotation (e.g. Helm chart-only sources) are skipped.
+func FilterByFiles(apps []argoappv1.Application, files []string) []argoappv1.Application {
+	if len(files) == 0 {
+		return apps
+	}
+	items := []argoappv1.Application{}
+	for i := range apps {
+		for _, source := range apps[i].Spec.GetSources() {
+			_, hasAnnotation := apps[i].Annotations[argoappv1.AnnotationKeyManifestGeneratePaths]
```
Comment (ppapapetrou76) [inline util/argo/argo.go]: I think this part is already checked in `GetSourceRefreshPaths` - please double check and remove any duplicate checks to simplify code

**Review by ppapapetrou76 — COMMENTED** [review]
The PR correctly updates FilterByPath to iterate all sources, but leaves FilterByRepo and FilterByRepoP broken for multi-source apps.

Please consider updating all FilterXYZ methods used in the `app list`

File: util/argo/argo_test.go [hunk 2]
```diff
@@ -789,6 +789,112 @@ func TestFilterByPath(t *testing.T) {
...
+		assert.Len(t, res, 1)
+		assert.Equal(t, appWithPath, res[0])
+	})
+
+	t.Run("Matches app with manifest-generate-paths annotation", func(t *testing.T) {
+		all := []argoappv1.Application{appWithAnnotation, appUnrelated}
+		res := FilterByFiles(all, []string{"apps/annotated/values.yaml"})
+		assert.Len(t, res, 1)
+		assert.Equal(t, appWithAnnotation, res[0])
+	})
```
Comment (nitishfy) [inline util/argo/argo_test.go:968]: It doesn't look like the test is actually testing what it is supposed to test. If you strip the annotation, it still passes i believe.

File: util/argo/argo.go [hunk 3]
```diff
@@ -177,15 +178,46 @@ func FilterByRepoP(apps []*argoappv1.Application, repo string) []*argoappv1.Appl
...
+			if source.Path == "" && !hasAnnotation {
```
Comment (nitishfy) [inline util/argo/argo.go]: a chart source with the annotation set to "." still hits the `AppFilesHaveChanged(["."], files)` path and it becomes `/.`, which `EnforceToCurrentRoot` treats as the repo root, so every file change matches. This is probably a configuration users won't actually have, but worth either documenting or safeguarding it more.

File: cmd/argocd/commands/app.go [hunk 4]
```diff
@@ -1471,6 +1475,7 @@ func NewApplicationListCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 	command.Flags().StringVarP(&appNamespace, "app-namespace", "N", "", "Only list applications in namespace")
 	command.Flags().StringVarP(&cluster, "cluster", "c", "", "List apps by cluster name or url")
 	command.Flags().StringVarP(&path, "path", "P", "", "List apps by path")
+	command.Flags().StringArrayVarP(&files, "file", "f", []string{}, "List apps affected by the given changed files (evaluates source path and manifest-generate-paths annotation)")
```
Comment (nitishfy) [inline cmd/argocd/commands/app.go]: The other argocd commands use `-f` mostly for file input, eg. `argocd app create -f app.yaml` but here the shorthand stands for filter by changed file. This is okay but it's confusing imo. I think we should have a different shorthand.

**Review by nitishfy — DISMISSED** [review]
Left some comments and need more tests.

**Comment by Prayasvadlakonda** [comment]
Pushed an update addressing the review feedback:

- **Removed the duplicate annotation lookup** in `FilterByFiles` — whether a source has annotation-derived refresh paths is now determined solely via `GetSourceRefreshPaths`'s return value, instead of separately re-checking the annotation.
- **Documented the `manifest-generate-paths: "."` edge case** — it resolves to the repo root, so it matches any changed file. This mirrors the annotation's existing meaning elsewhere, so I left it as documented behavior rather than special-casing it, but flagged it in the doc comment since it can make `--file` a no-op for that source.
- **Rewrote the annotation-match test** to use a chart-only source (no `Path`), so the match can only come from the annotation rather than falling back to the source path — the previous test passed even with the annotation stripped, which was the bug you flagged.
- **Fixed `FilterByRepo`/`FilterByRepoP`** to check all of a multi-source app's sources (same pattern as `FilterByPath`), instead of only `GetSource()` (first source). Added multi-source test cases for both.
- **Dropped the `-f` shorthand** on `--file` since `-f` already means "file input" elsewhere (`app create -f`, `project create -f`). `--file` now has no shorthand to avoid the collision.

Also rebased onto `master` and fixed the DCO check. All local builds/tests pass; CI is running now.

**Review by nitishfy — COMMENTED** [review]
Is this PR being worked upon?

**Review by ppapapetrou76 — APPROVED** [review]
LGTM 🏅  - thanks for addressing my comments

**Comment by Prayasvadlakonda** [comment]
@nitishfy yes, sorry for the slow turnaround — I pushed an update addressing all the review feedback back on July 1st (see [this comment](https://github.com/argoproj/argo-cd/pull/27866#issuecomment-4860666290) for the full list of changes): removed the duplicate annotation check, fixed `FilterByRepo`/`FilterByRepoP` for multi-source apps, rewrote the weak annotation test, documented the `"."` edge case, and dropped the `-f` shorthand collision. @ppapapetrou76 took another look and approved. Would appreciate a re-review when you have a chance — happy to make further changes if anything's still off.
